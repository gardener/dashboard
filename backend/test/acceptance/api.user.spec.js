//
// Copyright (c) 2019 by SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

'use strict'

module.exports = function ({ agent, sandbox, k8s, auth }) {
  /* eslint no-unused-expressions: 0 */
  const name = 'bar'
  const username = `${name}@example.org`
  const id = username
  const user = auth.createUser({ id })

  it('should return information about the user', async function () {
    const bearer = await user.bearer
    k8s.stub.getPrivileges({ bearer })
    const res = await agent
      .get('/api/user/privileges')
      .set('cookie', await user.cookie)

    expect(res).to.have.status(200)
    expect(res).to.be.json
    expect(res.body).to.eql({
      isAdmin: false
    })
  })

  it('should return the bearer token of the user', async function () {
    const bearer = await user.bearer
    const res = await agent
      .get('/api/user/token')
      .set('cookie', await user.cookie)

    expect(res).to.have.status(200)
    expect(res).to.be.json
    expect(res.body.token).to.equal(bearer)
  })

  it('should return selfsubjectrules for the user', async function () {
    const bearer = await user.bearer
    const project = 'foo'
    const namespace = `garden-${project}`
    k8s.stub.getSelfSubjectRulesReview({ bearer, namespace })
    const res = await agent
      .post('/api/user/subjectrules/')
      .set('cookie', await user.cookie)
      .send({ namespace })

    expect(res).to.have.status(200)
    expect(res).to.be.json
    expect(res.body).to.have.property('resourceRules')
    expect(res.body).to.have.property('nonResourceRules')
    expect(res.body).to.have.property('incomplete')
    expect(res.body.resourceRules.length).to.equal(2)
  })
}
