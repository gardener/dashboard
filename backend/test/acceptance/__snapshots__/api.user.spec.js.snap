// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api user should return information about the user 1`] = `
Array [
  Array [
    Object {
      ":authority": "kubernetes:6443",
      ":method": "post",
      ":path": "/apis/authorization.k8s.io/v1/selfsubjectaccessreviews",
      ":scheme": "https",
      "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImJhckBleGFtcGxlLm9yZyIsImlhdCI6MTU3NzgzNjgwMCwiYXVkIjpbImdhcmRlbmVyIl0sImV4cCI6MzE1NTcxNjgwMCwianRpIjoianRpIn0.7WKy0sNVkJzIqh3QJIF1zk3QjzwFe_zMTv8PmnOCsxg",
    },
    Object {
      "apiVersion": "authorization.k8s.io/v1",
      "kind": "SelfSubjectAccessReview",
      "spec": Object {
        "nonResourceAttributes": undefined,
        "resourceAttributes": Object {
          "group": "",
          "resource": "secrets",
          "verb": "get",
        },
      },
    },
  ],
]
`;

exports[`api user should return information about the user 2`] = `
Object {
  "isAdmin": false,
}
`;

exports[`api user should return selfsubjectrules for the user 1`] = `
Array [
  Array [
    Object {
      ":authority": "kubernetes:6443",
      ":method": "post",
      ":path": "/apis/authorization.k8s.io/v1/selfsubjectrulesreviews",
      ":scheme": "https",
      "authorization": "Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImJhckBleGFtcGxlLm9yZyIsImlhdCI6MTU3NzgzNjgwMCwiYXVkIjpbImdhcmRlbmVyIl0sImV4cCI6MzE1NTcxNjgwMCwianRpIjoianRpIn0.7WKy0sNVkJzIqh3QJIF1zk3QjzwFe_zMTv8PmnOCsxg",
    },
    Object {
      "apiVersion": "authorization.k8s.io/v1",
      "kind": "SelfSubjectRulesReview",
      "spec": Object {
        "namespace": "garden-foo",
      },
    },
  ],
]
`;

exports[`api user should return selfsubjectrules for the user 2`] = `
Object {
  "incomplete": false,
  "nonResourceRules": Array [],
  "resourceRules": Array [
    Object {
      "apiGroups": Array [
        "core.gardener.cloud",
      ],
      "resourceName": Array [
        "foo",
      ],
      "resources": Array [
        "projects",
      ],
      "verbs": Array [
        "get",
      ],
    },
    Object {
      "apiGroups": Array [
        "core.gardener.cloud",
      ],
      "resources": Array [
        "projects",
      ],
      "verbs": Array [
        "create",
      ],
    },
  ],
}
`;

exports[`api user should return the bearer token of the user 1`] = `
Object {
  "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImJhckBleGFtcGxlLm9yZyIsImlhdCI6MTU3NzgzNjgwMCwiYXVkIjpbImdhcmRlbmVyIl0sImV4cCI6MzE1NTcxNjgwMCwianRpIjoianRpIn0.7WKy0sNVkJzIqh3QJIF1zk3QjzwFe_zMTv8PmnOCsxg",
}
`;

exports[`api user should return the kubeconfig data the user 1`] = `
Object {
  "certificateAuthorityData": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCkxpNHUKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQ==",
  "oidc": Object {
    "certificateAuthorityData": "LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCkxpNHUKLS0tLS1FTkQgQ0VSVElGSUNBVEUtLS0tLQ==",
    "clientId": "kube-kubectl",
    "clientSecret": "6b7562652d6b75626563746c2d736563726574",
    "extraScopes": Array [
      "email",
      "profile",
      "groups",
    ],
    "issuerUrl": "https://kubernetes:32001",
  },
  "server": "https://kubernetes.external.foo.bar",
}
`;
